### ISSUE TITLE
Big difference between Chap.4 Exercise 12 Solution and sklearn built-in function results

### ISSUE THREAD

**Issue by LittleCottage** [ageron/handson-ml2#359]
In Chapter 4, I am trying to understand the non-negligible gap between the coefficient estimation "Theta" (given by the solution) and the "intercept_" + "coef_" (given by running LogisticRegression). The former gives
```
array([[ 3.32094157, -0.6501102 , -2.99979416],
       [-1.1718465 ,  0.11706172,  0.10507543],
       [-0.70224261, -0.09527802,  1.4786383 ]])
```
while the latter gives
```
[  9.80485039   2.64065217 -12.44550256]
[[-2.37655394 -0.93923633]
 [ 0.06849808 -0.75771513]
 [ 2.30805586  1.69695147]]
```
The code I used to get the latter result is 
```
from sklearn.linear_model import LogisticRegression

softmax_reg = LogisticRegression(multi_class = 'multinomial', solver = 'lbfgs')
# We need to drop the "ones" column to fit in Scikit-Learn methods
X_train_sklearn = X_train[:, 1:]
softmax_reg.fit(X_train_sklearn, y_train)
print(softmax_reg.intercept_)
print(softmax_reg.coef_)

```

Except for the possible stochastic nature in the algorithm (which shouldn't lead to such big difference), what caused the gap above?

**Comment by ageron** [comment]
Thanks for your question. If you run this code:

```python
Theta = np.zeros([3, 3])
Theta[0] = softmax_reg.intercept_
Theta[1:] = softmax_reg.coef_.T
```

then run the last two cells in the notebook, you will find that the `LogisticRegression` solution and the Batch Gradient Descent solution are actually quite close:

**Batch Gradient Descent Solution**

![image](https://user-images.githubusercontent.com/76661/109550089-b24ef680-7b33-11eb-93e0-efe69809673d.png)

**LogisticRegression Solution**

![image](https://user-images.githubusercontent.com/76661/109550140-c2ff6c80-7b33-11eb-821b-64b9ecb49311.png)

And both solutions give exactly the same accuracy score: 0.9333333333333333.

The weights look quite different, but around the region of interest (i.e., where all the instances are), they produce roughly the same boundaries. It's a bit like 2x + 3 and 3x - 1: these are different equations, but around x=4 they are very similar.

So the stochastic nature of the algorithm is sufficient to explain the difference. Also, recall that we are using early stopping with Batch Gradient Descent, which may also introduce a little difference.

In the end, both models are good, and despite the appearance they are actually quite similar.

Hope this helps.

**Comment by LittleCottage** [comment]
Thank you very much for your excellent explanation. It makes great sense!
